docs/book/v2/migration.md

```diff
@@ -0,0 +1,83 @@
+# Migration from Version 1
+
+This document details changes made from version 1 of this package or from its implementation in
+[laminas-mvc](https://docs.laminas.dev/laminas-mvc/).
+If installed, it overrides the default behavior in laminas-mvc.
+
+## `'controller'` Route Array Key
+
+In addition to providing a `'middleware'` array key in the route defaults definition, the value of the `'controller'`
+array key has to be explicitly set to `Laminas\Mvc\Middleware\PipeSpec::class`:
+```php
+/* ... */
+'some-route' => [
+    'type' => Literal::class,
+    'options' => [
+        'route' => '/some-route',
+        'defaults' => [
+            'controller' => \Laminas\Mvc\Middleware\PipeSpec::class,
+            'middleware' => SomeRequestHandler::class,
+        ],
+    ],
+],
+```
+
+## `PipeSpec` Instance in `middleware` Definition
+
+If you used to pass an array of middlewares to the `middleware` array key to create a middleware pipe, it has to be
+rewritten to an instance of the `PipeSpec` object.
+
+Before:
+
+```php
+/* ... */
+'some-route' => [
+    'type' => Literal::class,
+    'options' => [
+        'route' => '/some-route',
+        'defaults' => [
+            'middleware' => [
+                SomeMiddleware::class,
+                SomeRequestHandler::class,
+            ],
+        ],
+    ],
+],
+```
+
+After:
+
+```php
+/* ... */
+'some-route' => [
+    'type' => Literal::class,
+    'options' => [
+        'route' => '/some-route',
+        'defaults' => [
+            'controller' => \Laminas\Mvc\Middleware\PipeSpec::class,
+            'middleware' => new \Laminas\Mvc\Middleware\PipeSpec(
+                SomeMiddleware::class,
+                SomeRequestHandler::class
+            ),
+        ],
+    ],
+],
+```
+
+## PSR-15 Middleware
+
+This package now supports only PSR-15 interfaces. Support of `http-interop/http-middleware` and `callable` middleware
+has been dropped.
+
+You have to rewrite all your existing handlers and middleware to implement `Psr\Http\Server\RequestHandlerInterface`
+or `Psr\Http\Server\MiddlewareInterface`, respectively.
```

Comment:
... "or decorate them using one of the [decorators provided by Stratigility](https://docs.laminas.dev/laminas-stratigility/v3/creating-middleware/)."